### PR TITLE
treewide: Improve docker container build, CI caching and python dependencies

### PR DESCRIPTION
**Pull request by fischeti**
## Docker container

* Use the official Verilator container as final base container, which has `verilator` prebuilt
* Improve docker build stages for better caching
* Clean up apt installs and getting rid of `apt-requirements.txt`
* Bump python to `3.11`
* Bump Verilator to `5.020`

## CI

* Build Docker container on every branch as part of the `ci.yml` workflow. Before, the docker container had to be manually changed and triggered on the working branch if the Docker container needed modification.
* Cache docker build (~3min if cached, ~15min for clean build)
* Cache Verilator simulation build if file dependencies (generated from Makefiles) remain unchanged (saves ~20min)
* Disable redundant runs for `pull_request` if the pull request is from an internal contributor.

## Python dependencies

* Consolidate all dependencies into the `pyproject.toml` file, replaces all `requirements.txt` files.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -12,8 +12,52 @@ authors = [
...
+    "pytest",
+    "pyyaml",
+    "scikit-learn",
+    "tabulate",
+    "termcolor",
+    "yamllint",
+    # DNN lib dependencies.
+    "numpy",
+    "torch",
+    "json5"
```
[the remark this answers is not in the document]
Comment (fischeti) [reply pyproject.toml:47]: Yeah, you are probably right. There is not a big reason to separate them. I will merge all of them.

And yes, `json5` seems to be duplicate ;)

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -16,14 +16,6 @@ variables:
...
   source iis-setup.sh
 
-##############
-# Build docs #
-##############
-
-docs:
-  script:
-    - make docs
-
```
[the remark this answers is not in the document]
Comment (fischeti) [reply .gitlab-ci.yml:26]: Will revert, when I merge the dependencies and the `docs` dependencies are not optional anymore

File: .github/workflows/ci.yml
[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/ci.yml:1]: I thought about that. What is a bit annoying is that without the `pip install .` in the CI, the user will have to build a new docker container everytime there is a change in the package. But I don't have a strong opinion, what do you think?

[the remark this answers is not in the document]
Comment (colluca) [reply .github/workflows/ci.yml:1]: I agree that this would be pretty annoying with the current CI scripts.

We could probably improve it in terms of automation, such that we don't have to manually hack the deployment of a new container every time. But unless you have time to do this, I would maybe postpone this, and just have the `pip install .` in the CI scripts and remove it altogether from the container.

[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/ci.yml:1]: I will try a bit to improve the container build, if I fail I will keep the `pip install .` in the CI scripts

**Comment by fischeti** [comment]
Undrafting this again after quite a lot of changes. The PR now touches a lot more than just consolidating python dependencies (see updated description).

File: util/container/Dockerfile [hunk 3]
```diff
@@ -4,174 +4,127 @@
...
+
+# 2. Stage: Filter python files to improve caching of the final container
+FROM ubuntu:${UBUNTU_VERSION} AS python_filter
+
+# Copy all files in `sw` and `util` and remove all non-python files
+COPY sw ./sw
+COPY util ./util
+
+RUN find sw -type f ! -name "*.py" -delete && \
+    find util -type f ! -name "*.py" -delete
```
Comment (colluca) [inline util/container/Dockerfile:75]: If I understand correctly, any change in `sw` or `util` would cause it to rebuild instead of fetch from the cache, right?

In this case, although there is no glob support in Docker, this [thread](https://stackoverflow.com/questions/35670907/docker-copy-with-file-globbing) specifies how one could use a `.dockerignore` file to achieve this purpose. We could maybe use this to make it more fine-grained.

What do you think about this? I would leave the decision up to you, we could even come back to this later on.

Edit:
Going through the file I now understand you are using this stage to implement solution 3 in the StackOverflow thread linked above. Maybe we can extend the comment a bit so we don't forget about this in the future.

File: util/container/Dockerfile [hunk 4]
```diff
@@ -4,174 +4,127 @@
...
+ARG UBUNTU_VERSION=22.04
+
+# 1. Stage: Install additional IIS tools
+FROM ubuntu:${UBUNTU_VERSION} AS builder
 ARG BENDER_VERSION=0.27.1
-ARG SPIKE_DASM_VERSION=0.1.0
-ARG VERILATOR_VERSION=5.006
 ARG DOXYGEN_VERSION=1.12.0
-# Run dpkg without interactive dialogue
-ARG DEBIAN_FRONTEND=noninteractive
```
Comment (colluca) [inline util/container/Dockerfile:16]: I believe it would be better to keep the `DEBIAN_FRONTEND` variable in every stage where we use `apt`

File: util/container/Dockerfile [hunk 5]
```diff
@@ -4,174 +4,127 @@
...
-        wget \
-        build-essential \
-        git && \
-    apt-get clean ; \
+# Install Python
+RUN apt-add-repository ppa:deadsnakes/ppa && \
+    apt update && \
+    apt install -y --no-install-recommends \
+        python${PYTHON_VERSION}-venv && \
+    apt clean && \
```
Comment (colluca) [inline util/container/Dockerfile:102]: We can potentially merge these two `RUN` commands right?

File: util/container/Dockerfile [hunk 6]
```diff
@@ -4,174 +4,127 @@
...
-# Add Verilator to PATH
-ENV PATH "/tools/verilator/bin:${PATH}"
-ENV VLT_ROOT "/tools/verilator/share/verilator/"
-
-# Set locale to UTF-8, required because Python 3.6 defaults on ASCII encoding.
-# See https://click.palletsprojects.com/en/8.1.x/unicode-support/
-ENV LC_ALL "C.UTF-8"
-ENV LANG   "C.UTF-8"
+ENV PATH="/tools/bin:${PATH}"
+ENV PATH="/tools/riscv-llvm/bin:${PATH}"
```
Comment (colluca) [inline util/container/Dockerfile:130]: Is it necessary to add `LLVM_BINROOT` to the `PATH`? Everything should work even without, as we don't rely on this on IIS machines.

**Review by colluca — CHANGES_REQUESTED** [review]
Thanks for the extensive efforts @fischeti!

It looks very good, just have a few questions/comments.

File: .github/workflows/build-docker.yml
[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/build-docker.yml:1]: It would actually be possible to [depend on another workflow](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#running-a-workflow-based-on-the-conclusion-of-another-workflow), but it was easier to just move it to `ci.yml`

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -42,20 +84,34 @@ jobs:
...
+      - name: Generate the RTL
+        working-directory: target/snitch_cluster
+        run: |
+          make CFG_OVERRIDE=cfg/github-ci.hjson rtl
+      - name: Set up cache for Verilator build
+        id: verilator-cache
+        uses: actions/cache@v3
+        with:
+          path: target/snitch_cluster/bin
+          key: verilator-${{ hashFiles('Bender.lock', 'hw/**/*.sv', 'target/**/*.sv') }}
```
[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/ci.yml:110]: Yes, I know. This was actually  my first approach with a `vlt-flist` make target. But the `hashFiles` function is a bit annyoing, since it requires comma-separated files, which I didn't have the energy anymore to debug😄 

If you want to pick it up, this is the commit https://github.com/pulp-platform/snitch_cluster/pull/192/commits/3dfdbfc268a143e2807037998036daf9d7d4155c where I reverted to the simpler hashing of files

File: .github/workflows/lint.yml [hunk 8]
```diff
@@ -42,27 +48,29 @@ jobs:
...
-        run: pip install .
+      - name: Install bender
+        uses: pulp-platform/pulp-actions/bender-install@v2
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.10'
+          cache: 'pip'
+      - name: Install pip dependencies
+        run: |
+          pip install .
```
[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/lint.yml:66]: The docker build for the branch is in the `ci.yml` workflow, and I wanted to prevent workflow dependencies. But fetching the container from `main` is probably also fine, so we can also revert this.

File: iis-setup.sh [hunk 9]
```diff
@@ -8,19 +8,17 @@ export BENDER=bender-0.27.1
...
 export CXX=g++-9.2.0
 export VCS_SEPP=vcs-2020.12
-export VERILATOR_SEPP=verilator-5.006
+export VERILATOR_SEPP=verilator-5.020
 export QUESTA_SEPP=questa-2022.3
 export LLVM_BINROOT=/usr/pack/riscv-1.0-kgf/pulp-llvm-0.12.0/bin
 
 # Create Python virtual environment with required packages
-/usr/local/anaconda3-2022.05/bin/python3 -m venv .venv
+/usr/local/anaconda3/bin/python -m venv .venv
```
[the remark this answers is not in the document]
Comment (fischeti) [reply iis-setup.sh:16]: True, in that case, we should use `/usr/local/anaconda3-2023.07/bin/python`, which is also a Python 3.11 version.

[the remark this answers is not in the document]
Comment (fischeti) [reply iis-setup.sh:16]: I will fix this quickly.

File: util/container/Dockerfile (hunk 3 above)
Comment (fischeti) [reply util/container/Dockerfile:75]: > If I understand correctly, any change in sw or util would cause it to rebuild instead of fetch from the cache, right?

Yes, exactly

> In this case, although there is no glob support in Docker, this [thread](https://stackoverflow.com/questions/35670907/docker-copy-with-file-globbing) specifies how one could use a .dockerignore file to achieve this purpose. We could maybe use this to make it more fine-grained.

Ah yes, `.dockerignore` might also be an option. I did not consider this, but is probably cleaner if it is possible.

> Edit:
Going through the file I now understand you are using this stage to implement solution 3 in the StackOverflow thread linked above. Maybe we can extend the comment a bit so we don't forget about this in the future.

Yes, that seems to be the same solution ChatGPT gave me😁

File: util/container/Dockerfile (hunk 4 above)
Comment (fischeti) [reply util/container/Dockerfile:16]: It is only used for a single package. Otherwise, the container would not finish building since it expects an input from the user.

File: util/container/Dockerfile (hunk 5 above)
Comment (fischeti) [reply util/container/Dockerfile:102]: Yes, they could be merged I believe

File: util/container/Dockerfile (hunk 6 above)
Comment (fischeti) [reply util/container/Dockerfile:130]: I think I only added it for `clang-format`

File: .github/workflows/lint.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (colluca) [reply .github/workflows/lint.yml:66]: I'm mainly scared of having different dependencies in different places, e.g. the Python version here is already different from the one we use in the container :grin:

Is there a particular reason to avoid workflow dependencies? Or is just more complex?

[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/lint.yml:66]: It is just more complex/verbose. But I agree, it would be better to use the docker container. I will add a workflow dependency.

[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/lint.yml:66]: I reconsidered 😄 workflow dependencies are annyoing, since they apply to the entire workflow. So with the current workflows, we would have to wait for the entire `ci.yml` workflow to run before the `lint.yml` can start. Also, we would have to [filter out workflow failures](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#running-a-workflow-based-on-the-conclusion-of-another-workflow), which will then just show up as not run instead of failures.

Long story short, I am not a big fan of workflow dependencies🫠

File: util/container/Dockerfile (hunk 3 above)
Comment (fischeti) [reply util/container/Dockerfile:75]: A `.dockerignore` file would be much cleaner. I will implement this

Comment (fischeti) [reply util/container/Dockerfile]: implemented in https://github.com/pulp-platform/snitch_cluster/pull/192/commits/08f4840678fc0f30530b5de56799c6e2158d8972

File: .github/workflows/lint.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (colluca) [reply .github/workflows/lint.yml:66]: But regarding the first problem we can just move the Docker build to its own workflow as it was, and it is solved.

Regarding the second, I think the dependent job not failing is not a problem, as for it to be reported as not being run the Docker build would already have to fail. Or did I get it wrong?

File: util/container/Dockerfile (hunk 4 above)
Comment (colluca) [reply util/container/Dockerfile:16]: No, I mean we should not only use where it fails otherwise, but everywhere we use `apt`, as a guarantee that it will never fail, as a consequence of packages requiring user input.

Also, now that I look at this again, use of `apt` in scripts is discouraged, in favour of `apt-get`, see the note at https://ubuntu.com/server/docs/package-management#removing-packages.

**Comment by colluca** [comment]
I pushed commit 4f89a6f9fb4052b8cceaab57e1a5ab9d530d7795 to automate generation of the dependency list which determines the Verilator model build.

It is based on the [List Make Prerequisites Action](https://github.com/marketplace/actions/list-make-prerequisites-action), which parses Make's database output (from `make -pq`), to generate the dependency tree for any top-level target (in our case `bin/snitch_cluster.vsim`). It then lists all leaf nodes of the tree, i.e. all file dependencies the target ultimately depends on, and hashes their contents.

The method is still not 100% correct, e.g. if `FESVR_VERSION` is changed it will not trigger a cache miss. A solution would be to include the Makefiles themselves in the list of dependencies. The [List Make Prerequisites Action](https://github.com/marketplace/actions/list-make-prerequisites-action) could be extended to this end, but probably also the Makefiles would have to be split up to mix the least amount of logic, to not trigger unnecessary builds on every Makefile edit.

Nonetheless, the method is at least as precise as the previously proposed method, so I think we can continue with this solution, and eventually refine it in the future, when needed.

File: .github/workflows/lint.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (fischeti) [reply .github/workflows/lint.yml]: I tried to implement that, but the workflows weren't triggered and  saw that the documentation of [`workflow_run`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_run) says:

> Note: This event will only trigger a workflow run if the workflow file is on the default branch.

So, I believe workflow dependencies are not an option since this only works on the `main` branch for some reason

File: util/container/Dockerfile (hunk 4 above)
Comment (fischeti) [reply util/container/Dockerfile:16]: Addressed in https://github.com/pulp-platform/snitch_cluster/pull/192/commits/25fc2942865c0901c8dce6cb5223355e2ebb871e and https://github.com/pulp-platform/snitch_cluster/pull/192/commits/79a03dffed611c947287a06ed038a4f6e31d9424

File: .github/workflows/lint.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (colluca) [reply .github/workflows/lint.yml:66]: Oh, well if it's a Github limitation then we might as well put everything in a single workflow file, or leave it as you suggested before.

[the remark this answers is not in the document]
Comment (colluca) [reply .github/workflows/lint.yml]: I moved the `sources-up-to-date` job from the lint to the CI workflow, so we can encode the dependency to the docker build job and execute it on the Docker container.

**Review by colluca — APPROVED** [review]
